### PR TITLE
feat(tools): hint-based allowlist and MCP annotation hints

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Advances issue #607 by adding `hint:`-prefixed allowlist patterns and surfacing MCP tool safety annotations as ToolSpec hints.

- **`ToolSpec.hints: frozenset[str]`** — new field alongside the existing `ToolFunction.hints`; carries capability tags like `"read-only"`, `"destructive"`, `"idempotent"`, `"closed-world"`

- **`hint:` allowlist prefix** — agents can now write `tool_allowlist = ["hint:read-only", "shell"]` to auto-allow all tools tagged read-only, without enumerating every MCP function individually. Directly addresses the "lethal trifecta" safety concern in #607.

- **MCP annotation extraction** — `create_mcp_tools()` now maps `ToolAnnotations` from the MCP spec into `ToolSpec.hints` (`readOnlyHint` → `"read-only"`, `destructiveHint` → `"destructive"`, etc.)

- **Validation bypass for hint patterns** — `get_toolchain()` and `init_tools()` skip name-lookup validation for `hint:` entries; `allowlist_contains_glob()` returns True when hints are present (suppresses the skipped-MCP-tools warning)

- **38 new tests** covering `is_hint_pattern`, `allowlist_contains_glob`, `matching_allowlist_tools`, `tool_matches_allowlist`, and `ToolSpec.hints`

The `ToolFunction` abstraction itself landed in #2880.

## Test plan

- [x] `uv run pytest tests/test_tools_base.py` — 119 passed
- [x] Pre-commit hooks pass (ruff format, mypy)

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
Fixed the `destructiveHint` check found by Greptile — changed `if ann.destructiveHint:` to `if ann.destructiveHint is not False:`. Per the MCP spec, `destructiveHint` should be assumed True when absent (unlike `readOnlyHint` which defaults to False), so a truthy check silently dropped the `"destructive"` tag for unannotated tools, making `hint:destructive` under-match. Matches how `openWorldHint` is already handled in the same block (also `is not False`).

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Two remaining Greptile findings addressed in b4967930:

1. **`readOnlyHint`/`destructiveHint` contradiction** — when `readOnlyHint=True` and `destructiveHint` is absent (None, which per MCP spec defaults to True), the old code tagged the tool as both `"read-only"` and `"destructive"`, which is contradictory. Fixed with `if ann.destructiveHint is not False and not ann.readOnlyHint:` — a read-only tool cannot be destructive regardless of the missing-means-True default. Updated the test to assert the corrected hint set.

2. **Subagent `tool.hints` not forwarded** — `subagent/execution.py` line 167 called `tool_matches_allowlist(tool.name, tool_allowlist)` without passing `tool.hints`, so `hint:` allowlist entries silently matched nothing in subagent profiles. Fixed by passing the third argument.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Fixed the no-extras CI failure in `aec3165d`: `test_hint_allowlist_filters_subagent_tools` was using `patch("gptme.chat.chat")` which fails when the module hasn't been pre-imported — `gptme/__init__.py`'s lazy `__getattr__` caches the `chat` *function* under `gptme.chat`, so `getattr(gptme, 'chat')` returns the function and `patch` then fails looking for a `.chat` attribute on it.

Fixed to match the pattern already used by 4 other tests in the file: `import gptme.chat` explicitly first, then `patch.object(sys.modules["gptme.chat"], "chat")`. CI running.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Fixed the last remaining Greptile finding in `19daab37`:

**`hint:` patterns incorrectly flagged as "unknown tools"** — the set-comprehension at `execution.py:152-162` that builds the `unknown` set didn't skip `hint:` entries, so a valid `hint:read-only` pattern that matched no *currently-loaded* tools (e.g. because MCP servers haven't connected yet) would trigger `logger.warning("Profile '%s' references unknown tools: hint:read-only ...")` — factually wrong for a capability-tag filter.

Fix: added `not is_hint_pattern(pattern)` guard to the comprehension. Hint patterns are always valid; only name-glob entries can be misspelt.

Also removed the `matching_allowlist_tools` mock from `test_hint_allowlist_filters_subagent_tools` (which was there to suppress the false-positive warning) and added an explicit assertion that `hint:read-only` never appears in warning calls, so the fixed path is now actually tested.

File: gptme/tools/mcp_adapter.py [hunk 1]
```diff
@@ -229,6 +229,21 @@ def make_examples(
...
+                    ann = mcp_tool.annotations
+                    hint_set: set[str] = set()
+                    if ann.readOnlyHint:
+                        hint_set.add("read-only")
+                    if ann.destructiveHint:
+                        hint_set.add("destructive")
+                    if ann.idempotentHint:
+                        hint_set.add("idempotent")
+                    if ann.openWorldHint is False:
+                        hint_set.add("closed-world")
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/mcp_adapter.py:244]: Fixed in `c1cb19c0f7`: changed `if ann.destructiveHint:` to `if ann.destructiveHint is not False:` — matches the `openWorldHint` idiom already in the block.

File: gptme/tools/mcp_adapter.py [hunk 2]
```diff
@@ -229,6 +229,21 @@ def make_examples(
...
+                    hints = frozenset(hint_set)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/mcp_adapter.py:245]: Fixed in `f6d1e61cec`: added `tests/test_mcp_adapter.py` with coverage for `readOnlyHint=True`, `destructiveHint=True/False/None`, `idempotentHint`, `openWorldHint`, `annotations=None`, and the contradictory-dual-tag edge case.

File: gptme/tools/mcp_adapter.py [hunk 3]
```diff
@@ -229,6 +229,21 @@ def make_examples(
...
 
+                # Extract MCP ToolAnnotations into hint tags
+                hints: frozenset[str] = frozenset()
+                if mcp_tool.annotations:
+                    ann = mcp_tool.annotations
+                    hint_set: set[str] = set()
+                    if ann.readOnlyHint:
+                        hint_set.add("read-only")
+                    if ann.destructiveHint is not False:
+                        hint_set.add("destructive")
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/mcp_adapter.py:240]: Fixed in `b4967930e7`: the `destructiveHint` block now has an `and not ann.readOnlyHint` guard, so a `readOnlyHint=True` tool can't receive both `"read-only"` and `"destructive"` tags.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
PR is ready for merge: CI green, Greptile 5/5.

All findings from Greptile's review cycle have been addressed:
- `destructiveHint` default-to-true guard fixed (`is not False`)
- `readOnlyHint`/`destructiveHint` contradiction guard added
- MCP annotation → hint mapping tests added

One minor note flagged in the Greptile summary (not blocking): the JSON example in `docs/tools.rst` describing MCP annotation configuration format doesn't match an actual gptme config file format. Happy to fix it before merge if you'd prefer, otherwise it can be a follow-up.
